Enable user telemetry opt-in / opt-out on cli

miah opened the pull request:
This adds `--enable-telemetry` and `--no-enable-telemetry` to the cli. Currently telemetry is enabled by default.

I wanted to also add the ability to opt-out and specify additional configuration via our config file, but thats going to require another issue as its not as straight forward (our config file is versioned and making changes is currently non-trivial).

closes https://github.com/inspec/inspec/issues/3849

miah committed:
Expose configuration for telemetry as method

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Allow enabling or disabling telemetry from the cli

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
load our configuration and set telemetry status

Allow overriding config object

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Ensure the collector is aware of the config class

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Pass in mock config object to tests and validate telemetry_enabled?

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
When telemetry is disabled Inspec.record_telemetry_data doesn't record

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Add a missing doc

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah commented on test/unit/utils/telemetry/collector_test.rb:
```diff
@@ -40,4 +40,14 @@ def test_reset_singleton
+
+  def test_telemetry_enabled
+    @collector.load_config(Inspec::Config.mock(telemetry: true))
+    assert @collector.telemetry_enabled?
```
I feel like these tests are pretty silly since I'm just testing that the configuration got loaded and a True/False was set, is there a better way to do this?

zenspider replied:
I would `assert`/`refute` before the `load_config` to show state change. Otherwise, I think these tests are fine.

miah commented on lib/inspec/config.rb:
```diff
@@ -67,6 +67,10 @@ def diagnose
+    def telemetry_enabled?
+      final_options['telemetry']
+    end
+
```
I felt it was better to add an accessor here rather than our collector needing to know the format of our options hash and reaching inside.

clintoncwolfe replied:
I'm unsure about this. Should the config object have *general* knowledge of the  major subsystems of InSpec ? I think yes (it already know about train credentials, for example). 

However, asking specifically if the telemetry system is enabled means that when you add another config option - like endpoint URL - you'll need to add another method to the Config (or else scatter such methods between the Collector and Config).

So, I think if you asked the Config, "I don't care how you got it, but can you give me all the Telemetry options as a hash" would be fine - Collector doesn't have to care about Config's innards, and vice-versa.

miah commented on lib/inspec/utils/telemetry/collector.rb:
```diff
@@ -21,7 +27,11 @@ def add_data_series(data_series)
+      else
+        @enabled = @config.telemetry_enabled?
+      end
     end
```
I know this is a bit weird if `@config` is nil and thus the `telemetry_enabled?` method won't exist. This might be more problematic if InSpec is used as a 'library', but if the cli/config is loaded properly we should be good. Suggestions?

zenspider replied:
I'm not sure I understand your comment here. If `@config` is nil, that first line will fail calling `telemetry_enabled?` (unless we're doing hinky things on `NilClass`) and will raise a `NoMethodError`.

clintoncwolfe replied:
Inspec::Config.cached will still work if `inspec` is used as a library. Since that is the default, @config would only be nil if someone were actively trying to break it :-)

miah replied:
My only issue is that `Inspec::Config.cached` doesn't work 'right' if you are testing the collector class. This is why I'm passing in a mock config object.

zenspider reviewed (changes requested):
Minor change requested to 2 tests.

clintoncwolfe commented on lib/inspec/cli.rb:
```diff
@@ -39,6 +39,9 @@ class Inspec::InspecCLI < Inspec::BaseCLI
     desc: 'Disable loading all plugins that the user installed.'
 
+  class_option :telemetry, type: :boolean,
+    desc: 'Allow or disable telemetry', default: true
```
I'm pretty sure this should default to false - we're doing opt-in on other Chef products, IIRC...

clintoncwolfe replied:
Very weak suggestion, Consider naming this `--enable-telemetry` - I always like verb phrases

clintoncwolfe commented on lib/inspec/utils/telemetry/collector.rb:
```diff
@@ -8,7 +9,12 @@ class Collector
 
```
I suggest exposing @config via attr_reader, so we avoid the ivar accesses (which are vulnerable to typos)

clintoncwolfe reviewed (changes requested):
Thanks! I have a couple of changes I'd like:
 * Pretty sure this should be opt-in - let's verify with other teams, maybe Mike
 * I think the Telemetry-Config interface should know less about each other - see comment. A Hash would probably be fine for an interface.

miah committed:
cli -> `--enable-telemetry` and default to false

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Gather all telemetry options and return them

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Handle load_config better

private method to minimize external interface
default enabled to false if unable to find setting in configuration
add reader for config

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
condense true/false enabled tests

Because of the way the test system loads I cannot guarantee that
`Inspec::Config.cached` won't be nil. I have to always pass in a mock
config object with our setting.
Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
rubocop

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Update Inspec::Config.mock in collector global methods

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Ensure telemetry can be disabled

The #disable_telemetry method now toggles @telemetry_toggle_off.
The @telemetry_toggle_off is defaulted to false

The #telemetry_enabled? method now checks @telemetry_toggle_off
and we generally check the configuration. We do not flip
@telemetry_toggle_off based on the content of the configuration.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Update Collector#reset to clear telemetry_toggled_off

Updated the name of #reset -> #reset! as we are modifying state of a
singeton.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

zenspider commented on lib/inspec/utils/telemetry/collector.rb:
```diff
@@ -17,17 +26,20 @@ def add_data_series(data_series)
+      if @telemetry_toggled_off
+        false
+      else
+        config_telemetry_options.fetch('enable_telemetry', false)
```
I usually like my booleans to be positively oriented, because logical statements like this usually throw me for a bit of a De Morgan-y loop mentally. 

Also, if statements that have a return value of true/false in them is usually a sign that the logic can be cleaned up into an or/and expression... but see point #1. 😛

miah replied:
Ok. I'll give this one some more thought and see how I can make it better.

zenspider commented on lib/inspec/utils/telemetry/collector.rb:
```diff
@@ -52,9 +64,18 @@ def find_or_create_data_series(name)
+    # Reset telemetry toggle
     # @return [True]
-    def reset
+    def reset!
```
good

zenspider commented on lib/inspec/utils/telemetry/collector.rb:
```diff
@@ -52,9 +64,18 @@ def find_or_create_data_series(name)
+    private
+
+    # Minimize exposure of Inspec::Config interface
+    def config_telemetry_options
```
good!

zenspider commented on test/unit/utils/telemetry/collector_test.rb:
```diff
@@ -37,7 +37,24 @@ def test_find_or_create_data_series
     @collector.add_data_series(data_series)
-    @collector.reset
+    @collector.reset!
     assert_equal 0, @collector.list_data_series.count
```
Not part of your PR but `assert_empty` is more expressive here.

zenspider reviewed (approved):
Some minor feedback but looks good!

clintoncwolfe reviewed (approved):
I'm happy with it - thanks!
